lib/urlwatch/jobs.py

```diff
@@ -361,14 +361,78 @@ class BrowserJob(Job):
     __kind__ = 'browser'
 
     __required__ = ('navigate',)
+    __optional__ = ('cookies', 'data', 'ssl_no_verify', 'ignore_cached', 'http_proxy', 'https_proxy',
+                    'headers', 'ignore_connection_errors', 'ignore_http_error_codes', 'encoding', 'timeout',
+                    'ignore_timeout_errors', 'ignore_too_many_redirects')
 
     LOCATION_IS_URL = True
 
     def get_location(self):
         return self.navigate
 
     def retrieve(self, job_state):
+        headers = {
+            'User-agent': urlwatch.__user_agent__,
+        }
+
+        proxies = {
+            'http': os.getenv('HTTP_PROXY'),
+            'https': os.getenv('HTTPS_PROXY'),
+        }
+
+        if job_state.etag is not None:
+            headers['If-None-Match'] = job_state.etag
+
+        if job_state.timestamp is not None:
+            headers['If-Modified-Since'] = email.utils.formatdate(job_state.timestamp)
+
+        if self.ignore_cached or job_state.tries > 0:
+            headers['If-None-Match'] = None
+            headers['If-Modified-Since'] = email.utils.formatdate(0)
+            headers['Cache-Control'] = 'max-age=172800'
+            headers['Expires'] = email.utils.formatdate()
+
+        if self.http_proxy is not None:
+            proxies['http'] = self.http_proxy
+        if self.https_proxy is not None:
+            proxies['https'] = self.https_proxy
+
+        file_scheme = 'file://'
+        if self.navigate.startswith(file_scheme):
+            logger.info('Using local filesystem (%s URI scheme)', file_scheme)
+            from requests_html import HTML
+            return HTML(open(self.navigate[len(file_scheme):], 'rt').read())
+
+        if self.headers:
+            self.add_custom_headers(headers)
+
+        if self.timeout is None:
+            # default timeout
+            timeout = 60
+        elif self.timeout == 0:
+            # never timeout
+            timeout = None
+        else:
+            timeout = self.timeout
+
         from requests_html import HTMLSession
         session = HTMLSession()
-        response = session.get(self.navigate)
+        response = session.get(url=self.navigate,
+                               data=self.data,
+                               headers=headers,
+                               verify=(not self.ssl_no_verify),
+                               cookies=self.cookies,
+                               proxies=proxies,
+                               timeout=timeout)
         return response.html.html
+
+    def add_custom_headers(self, headers):
```

Comment:
Should the adding of custom headers be a separate PR? Can you add some unit tests for verifying that it works and does the right thing (replace keys with a different case, etc..)?

Also, there is a `CaseInsensitiveDict` in `requests.structures` that we could use, since we anyway depend on `requests`.

Reply:
No changes were made to the code copied from the URL job:
https://github.com/thp/urlwatch/blob/4654eb984dae6e1d7734fefcb3ae976f2d3d0843/lib/urlwatch/jobs.py#L321-L322